server/src/main/java/org/elasticsearch/search/aggregations/BucketOrder.java

```diff
@@ -98,16 +101,39 @@ public static BucketOrder compound(BucketOrder... orders) {
     }
 
     /**
-     * @return A comparator for the bucket based on the given aggregator. The comparator is used in two phases:
-     * <p>
-     * - aggregation phase, where each shard builds a list of buckets to be sent to the coordinating node.
-     * In this phase, the passed in aggregator will be the aggregator that aggregates the buckets on the
-     * shard level.
+     * Validate an aggregation against an aggregator.
+     */
+    public final void validate(Aggregator aggregator) {
+        /*
+         * Building partiallyBuiltBucketComparator and throwing it away is enough
+         * to validate this order because doing so checks all of the appropriate
+         * paths.
+         */
+        partiallyBuiltBucketComparator(null, aggregator);
+    }
+
+    /**
+     * A builds comparator comparing buckets partially built buckets by
+     * delegating comparison of the results of any "child" aggregations to
+     * the provided {@linkplain Aggregator}.
      * <p>
-     * - reduce phase, where the coordinating node gathers all the buckets from all the shards and reduces them
-     * to a final bucket list. In this case, the passed in aggregator will be {@code null}.
+     * Warning: This is fairly difficult to use and impossible to use cleanly.
+     * In addition, this exists primarily to return the "top n" buckets based
+     * on the results of a sub aggregation. The trouble is that could end up
+     * throwing away buckets on the data nodes that should ultimately be kept
+     * after reducing all of the results. If you know that this is coming it
+     * is fine, but most folks that use "generic" sorts don't. In other words:
+     * before you use this method think super duper hard if you want to have
+     * these kinds of issues. The terms agg does an folks get into trouble
+     * with it all the time.
+     * </p> 
+     */
+    public abstract <T extends Bucket> Comparator<T> partiallyBuiltBucketComparator(ToLongFunction<T> ordinalReader, Aggregator aggregator);
```

Comment:
Hmm, despite the warning it makes me a bit uncomfortable exposing this everywhere since only Terms needs it.  Perhaps we could have some kind of "OrderBySubaggWrapper" that Terms agg wraps over the `order` which implements this functionality? 

I'm not sure if that's possible (or a good idea), so I'm ok leaving this method + warning if we can't find a way to hide it.

Reply:
(I realize this sorta goes back to how it was before, but it at least avoids a special-case instanceof in ordering classes :), and retains all the other nice benefits of this refactor )

Reply:
The trouble with that is that validating the sort just calls this method. That prevents us from having all that "mirrored" validation code. I could make this method protected. What would you think of that?

Reply:
I feel like that'd be fairly tricky though.

Reply:
Hmm I see, and there's really no good way to currently validate without just running through the order components.

Ok, we can go with the current setup.  The javadocs are sufficiently scary, and we can keep it in mind if we ever find an opportunity to refactor and hide the method.

Reply:
So I've been poking around the PR a little, and I remembered this issue regarding validation: https://github.com/elastic/elasticsearch/issues/20003

E.g. the "validation" we do today is pretty poor, as it happens way after we parse the aggregations.  But the builders _do_ have a full list of all the aggregation builders (that's how pipeline aggs validate their `buckets_path`, albeit that system has it's own issues).

I wonder if we could -- in a followup PR -- make validation a parse-time issue, and then it would potentially allow us to remove validation from the Ordering classes and then also hide the terms-specific details?  WDYT?  Or are there other details that still make it difficult?

Reply:
That'd certainly make this something we could drop! I'd be happy to do that in a follow up. Reading it has got me thinking about ways we can do this without having so many "paths".

Reply:
Coolio.  A battle for a different day then!